synapse/util/caches/stream_change_cache.py

```diff
@@ -213,16 +213,17 @@ def has_any_entity_changed(self, stream_pos: int) -> bool:
         """
         assert isinstance(stream_pos, int)
 
-        if not self._cache:
-            # If the cache is empty, nothing can have changed.
-            return False
-
         # _cache is not valid at or before the earliest known stream position, so
         # return that an entity has changed.
```

Comment:
We should add a comment here that the ordering is important. Something like:

> It's important that we check stream position for changes first before we do the early-return check against the cache below. If the stream position is at/earlier than the earliest known stream position, but the cache is empty we should still consider entities to have changed.

Reply:
@clokep Thoughts? Picking this up?

Reply:
I don't think this is necessary -- the updated comments are clear and the tests assert this.

Reply:
@clokep I don't pick any of this information up from the comments. Only from your PR description.

It's great that we have tests for this though.